core/schema/module.go

```diff
@@ -321,182 +380,1162 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
+	return dagql.Null[core.Void](), modMeta.Self.Query.ServeModuleToMainClient(ctx, modMeta)
+}
+
+func (s *moduleSchema) currentTypeDefs(ctx context.Context, self *core.Query, _ struct{}) ([]*core.TypeDef, error) {
+	deps, err := self.CurrentServedDeps(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current module: %w", err)
+	}
+	return deps.TypeDefs(ctx)
+}
+
+func (s *moduleSchema) functionCallReturnValue(ctx context.Context, fnCall *core.FunctionCall, args struct {
+	Value core.JSON
+}) (dagql.Nullable[core.Void], error) {
+	// TODO: error out if caller is not coming from a module
+	return dagql.Null[core.Void](), fnCall.ReturnValue(ctx, args.Value)
+}
+
+func (s *moduleSchema) moduleWithDescription(ctx context.Context, mod *core.Module, args struct {
+	Description string
+}) (*core.Module, error) {
+	return mod.WithDescription(args.Description), nil
+}
+
+func (s *moduleSchema) moduleWithObject(ctx context.Context, mod *core.Module, args struct {
+	Object core.TypeDefID
 }) (_ *core.Module, rerr error) {
-	sourceDir, err := args.Directory.Load(ctx, s.dag)
+	def, err := args.Object.Load(ctx, s.dag)
 	if err != nil {
 		return nil, err
 	}
+	return mod.WithObject(ctx, def.Self)
+}
 
-	configPath, cfg, err := core.LoadModuleConfig(ctx, sourceDir.Self, args.Subpath)
+func (s *moduleSchema) moduleWithInterface(ctx context.Context, mod *core.Module, args struct {
+	Iface core.TypeDefID
+}) (_ *core.Module, rerr error) {
+	def, err := args.Iface.Load(ctx, s.dag)
 	if err != nil {
 		return nil, err
 	}
+	return mod.WithInterface(ctx, def.Self)
+}
+
+func (s *moduleSchema) currentModuleName(
+	ctx context.Context,
+	curMod *core.CurrentModule,
+	args struct{},
+) (string, error) {
+	return curMod.Module.Self.Name(), nil
+}
+
+func (s *moduleSchema) currentModuleSource(
+	ctx context.Context,
+	curMod *core.CurrentModule,
+	args struct{},
+) (inst dagql.Instance[*core.Directory], rerr error) {
+	rootDir := curMod.Module.Self.GeneratedSourceRootDirectory
+	subPath := curMod.Module.Self.GeneratedSourceSubpath
+	if subPath == "/" {
+		return rootDir, nil
+	}
+	var subDir dagql.Instance[*core.Directory]
+	err := s.dag.Select(ctx, rootDir, &subDir,
+		dagql.Selector{
+			Field: "directory",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(subPath)},
+			},
+		},
+	)
+	if err != nil {
+		return inst, fmt.Errorf("failed to get module source subdirectory: %w", err)
+	}
+	return subDir, nil
+}
+
+func (s *moduleSchema) currentModuleWorkdir(
+	ctx context.Context,
+	curMod *core.CurrentModule,
+	args struct {
+		Path string
+		core.CopyFilter
+	},
+) (inst dagql.Instance[*core.Directory], err error) {
+	if !filepath.IsLocal(args.Path) {
+		return inst, fmt.Errorf("workdir path %q escapes workdir", args.Path)
+	}
+	args.Path = filepath.Join(runtimeWorkdirPath, args.Path)
+	host := &hostSchema{srv: s.dag}
+	return host.directory(ctx, &core.Host{Query: curMod.Module.Self.Query}, args)
```

Comment:
Is it not possible to use `dagql.Select` here (and `workdirFile`)?

Reply:
It probably is possible to do that instead yeah, I went with this approach because it seemed simpler but I suppose it creates some entanglements with the `host` struct that could be annoying to deal with if we ever change it. Will update

Reply:
I made this update, but did remember that part of my motivation here was that in a follow-up I think we're going to block the `Host` API from the schema served to modules entirely, at which point I *think* I might need to go back to the previous implementation? 🤔 I'll just deal with it then, it's only the difference of a few lines

Reply:
Ah gotcha, makes sense, we'll see if we need to tweak it later